### PR TITLE
[SuperTextField] Fix layout with custom font height (Resolves #1864)

### LINKED ISSUE #1864: SuperTextField layout is wrong with custom text height
Video

https://github.com/superlistapp/super_editor/assets/96958/f516cca6-da66-4923-9f18-65b2babe58c0

You can see that the hint is aligned correctly while the actual text is not.

Modified snippet from `_single_line_demo.dart`

```dart
SuperDesktopTextField(
  key: _textKey,
  textController: _textFieldController,
  focusNode: _focusNode,
  decorationBuilder: (context, child) {
    return Container(
      decoration: BoxDecoration(
        borderRadius: BorderRadius.circular(4),
        border: Border.all(
          color: _focusNode!.hasFocus ? Colors.blue : Colors.grey.shade300,
          width: 1,
        ),
      ),
      child: child,
    );
  },
  hintBuilder: (context) {
    return const Text(
      'enter some text',
      style: TextStyle(
        color: Colors.grey,
        fontSize: 14,
        height: 2.0,
      ),
    );
  },
  hintBehavior: HintBehavior.displayHintUntilTextEntered,
  textStyleBuilder: (a) => demoTextStyleBuilder(a).copyWith(height: 2.0),
  minLines: 1,
  maxLines: 1,
),
```

### PR DESCRIPTION
**Pull request by angelosilvestre**
[SuperTextField] Fix layout with custom font height (Resolves #1864)

When using a custom font height, the text isn't aligned correctly:

https://github.com/superlistapp/super_editor/assets/96958/f516cca6-da66-4923-9f18-65b2babe58c0

With a bigger font height this issue is more noticeable: 

SCR-20240306-sfju

I found a Flutter issue about vertically aligning text with custom font heights: https://github.com/flutter/flutter/issues/72521. This issue was closed after a PR which adds a `leadingDistribution` property to `TextStyle` landed.

Apps need to configure the `TextStyle` with `TextLeadingDistribution.even`:

```dart
TextStyle textStyle = const TextStyle(
  color: Colors.black,
  fontSize: 14,
  leadingDistribution: TextLeadingDistribution.even,
);
```

With this change the text is centered in the caret, but not centered in the viewport:

SCR-20240306-shjc

This suggests that our viewport height calculation is wrong. Changing the calculation to use `textLayout.getHeightForCaret` instead of `textLayout.getLineHeightAtPosition` seems to fix the issue:

https://github.com/superlistapp/super_editor/assets/7597082/cdbe3a55-a738-4dc3-a5d0-e6d9d5f405e9

### CODE REVIEW

**Comment by angelosilvestre** [comment]
@knopp Could you please try this PR, setting the text style's `leadingDistribution` to `TextLeadingDistribution.even`, and see if it solves the issue for you?

**Comment by matthew-carroll** [comment]
@angelosilvestre in the example videos above, what's causing the height to be so large for such small text?

**Comment by angelosilvestre** [comment]
@matthew-carroll In the examples I used a font height of 6.0, just to make it more noticeable

File: super_editor/lib/src/super_textfield/desktop/desktop_textfield.dart [hunk 1]
```diff
@@ -363,7 +363,8 @@ class SuperDesktopTextFieldState extends State<SuperDesktopTextField> implements
     final textLayout = RenderSuperTextLayout.textLayoutFrom(_textKey);
     final lineHeight = _controller.text.text.isEmpty || textLayout == null
         ? 0.0
-        : textLayout.getLineHeightAtPosition(const TextPosition(offset: 0));
+        : textLayout.getHeightForCaret(const TextPosition(offset: 0)) ??
```
Comment (matthew-carroll) [inline super_editor/lib/src/super_textfield/desktop/desktop_textfield.dart:366]: This PR tries to avoid a changing height when we go from no text to some text right? If `getHeightForCaret()` returns `null`, aren't we right back at the original problem?

File: super_editor/test_goldens/super_textfield/super_textfield_font_height_test.dart [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+                        ),
+                      ),
+                      child: child,
+                    );
+                  },
+                  textStyleBuilder: (attributions) => const TextStyle(
+                    color: Colors.black,
+                    fontSize: 14,
+                    leadingDistribution: TextLeadingDistribution.even,
+                    height: 2.0,
```
Comment (matthew-carroll) [inline super_editor/test_goldens/super_textfield/super_textfield_font_height_test.dart:43]: To really make the test result clear, should we increase this `height` a lot?

File: super_editor/test_goldens/super_textfield/super_textfield_font_height_test.dart [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+import 'package:flutter/material.dart';
+import 'package:flutter_test/flutter_test.dart';
+import 'package:golden_toolkit/golden_toolkit.dart';
+import 'package:super_editor/src/super_textfield/desktop/desktop_textfield.dart';
+import 'package:super_editor/src/super_textfield/infrastructure/attributed_text_editing_controller.dart';
+
+import '../../test/super_textfield/super_textfield_robot.dart';
+import '../test_tools_goldens.dart';
+
+void main() {
```
Comment (matthew-carroll) [inline super_editor/test_goldens/super_textfield/super_textfield_font_height_test.dart:11]: Wasn't the original issue about a change in height between the hint text and the content text? If so, shouldn't we test for that somehow?

File: super_editor/lib/src/super_textfield/desktop/desktop_textfield.dart (hunk 1 above)
Comment (angelosilvestre) [reply super_editor/lib/src/super_textfield/desktop/desktop_textfield.dart:366]: If `getHeightForCaret` returns `true` then we would be back to the original problem, but it's not clear in which cases it can return `null`...

File: super_editor/test_goldens/super_textfield/super_textfield_font_height_test.dart (hunk 3 above)
Comment (angelosilvestre) [reply super_editor/test_goldens/super_textfield/super_textfield_font_height_test.dart:11]: Changing between empty and non-empty isn't particularly relevant for this issue. If the textfield is initialized with text or if it's initialized empty and then we type some text the result is the same. Also, during the tests `getLineHeightAtPosition` seems to always return the same height we estimate when the text is empty.

File: super_editor/test_goldens/super_textfield/super_textfield_font_height_test.dart (hunk 2 above)
Comment (angelosilvestre) [reply super_editor/test_goldens/super_textfield/super_textfield_font_height_test.dart]: Updated.

File: super_editor/lib/src/super_textfield/desktop/desktop_textfield.dart (hunk 1 above)
Comment (matthew-carroll) [reply super_editor/lib/src/super_textfield/desktop/desktop_textfield.dart:366]: Can you add a comment that explains that we don't expect the backup method to be called, but it's there because technically `getHeightForCaret` might fail?

Did you look in the Flutter source code to see why it might fail? If it's not clear in the source code, can you file a ticket with Flutter asking them to document the situations where it would fail?

File: super_editor/test_goldens/super_textfield/super_textfield_font_height_test.dart (hunk 3 above)
Comment (matthew-carroll) [reply super_editor/test_goldens/super_textfield/super_textfield_font_height_test.dart:11]: The video in the original ticket is of height changing between hint text and user-entered text.

**Comment by angelosilvestre** [comment]
@knopp Could you please share the `TextStyle`s from the text and the hint?

File: super_editor/lib/src/super_textfield/desktop/desktop_textfield.dart (hunk 1 above)
Comment (angelosilvestre) [reply super_editor/lib/src/super_textfield/desktop/desktop_textfield.dart:370]: I looked at the code and it seems it should never return `null`. I filed https://github.com/flutter/flutter/issues/145507 and mentioned that in the code.

File: super_editor/test_goldens/super_textfield/super_textfield_font_height_test.dart (hunk 3 above)
Comment (angelosilvestre) [reply super_editor/test_goldens/super_textfield/super_textfield_font_height_test.dart:11]: Yeah, but I don't think this is relevant to the golden test. If the implementation changes, the test would fail even without changing from hint to user-entered text.

**Comment by angelosilvestre** [comment]
@matthew-carroll It seems https://github.com/superlistapp/super_editor/pull/1839 caused a regression in `super_editor_markdown`. I think we should revert that.

**Comment by matthew-carroll** [comment]
@angelosilvestre I reverted that Markdown package upgrade.

**Comment by matthew-carroll** [comment]
@knopp it looks like Angelo is waiting for some updated details from your end.

**Comment by angelosilvestre** [comment]
@matthew-carroll knop answered in the original ticket saying that it worked for him.

**Comment by matthew-carroll** [comment]
@angelosilvestre does that mean this PR is ready for another review? Also, do you want to rebase and get the reverted Markdown code?

**Comment by angelosilvestre** [comment]
@matthew-carroll Yeah, it's ready for another review. I just rebased this branch.

File: super_editor/example/lib/demos/supertextfield/_single_line_demo.dart [hunk 4]
```diff
@@ -46,9 +47,9 @@ class _SingleLineTextFieldDemoState extends State<SingleLineTextFieldDemo> with
     _textFieldController
       ..selection = const TextSelection.collapsed(offset: 0)
       ..text = AttributedText();
-    _demoRobot
-      ..typeText(AttributedText('Hello World! This is a robot typing some text into a SuperTextField.'))
-      ..start();
+    // _demoRobot
```
Comment (matthew-carroll) [inline super_editor/example/lib/demos/supertextfield/_single_line_demo.dart:50]: Should this be brought back?

File: super_editor/example/lib/demos/supertextfield/_single_line_demo.dart [hunk 5]
```diff
@@ -76,35 +77,38 @@ class _SingleLineTextFieldDemoState extends State<SingleLineTextFieldDemo> with
...
-                            width: 1,
-                          ),
-                        ),
-                        child: child,
-                      );
-                    },
-                    hintBuilder: (context) {
-                      return const Text(
-                        'enter some text',
+                  child: ColoredBox(
```
Comment (matthew-carroll) [inline super_editor/example/lib/demos/supertextfield/_single_line_demo.dart:80]: Should all this be reverted? Or is this actually the desired demo UI?

File: super_editor/test_goldens/super_textfield/super_textfield_font_height_test.dart [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+  group('SuperTextField > single line > with custom font height', () {
+    testGoldensOnMac('centers text in viewport', (tester) async {
```
Comment (matthew-carroll) [inline super_editor/test_goldens/super_textfield/super_textfield_font_height_test.dart:13]: Can you say "vertically centers" so it's not ambiguous?

File: super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart [hunk 7]
```diff
@@ -1256,7 +1256,12 @@ class _RenderTextViewport extends RenderProxyBox {
     }
 
     final textLayout = _textKey.currentState!.textLayout;
-    double lineHeight = textLayout.getLineHeightAtPosition(const TextPosition(offset: 0));
+
+    // We don't expect getHeightForCaret to ever return null, but since its return type is nullable,
+    // we use getLineHeightAtPosition as a backup.
+    // More information in https://github.com/flutter/flutter/issues/145507.
+    double lineHeight = textLayout.getHeightForCaret(const TextPosition(offset: 0)) ??
```
Comment (matthew-carroll) [inline super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart:1263]: FYI - I noticed that on `master` there's a method that I think is `getFullHeightForCaret()` or something like that and Flutter made it non-null. I don't know if that's relevant here - I also don't know if maybe it now returns zero instead of `null`, but you might want to check if that's a better option for this.

File: super_editor/example/lib/demos/supertextfield/_single_line_demo.dart (hunk 4 above)
Comment (angelosilvestre) [reply super_editor/example/lib/demos/supertextfield/_single_line_demo.dart:50]: Yes, I shouldn't have commited this file.

File: super_editor/example/lib/demos/supertextfield/_single_line_demo.dart (hunk 5 above)
Comment (angelosilvestre) [reply super_editor/example/lib/demos/supertextfield/_single_line_demo.dart:80]: Yeah, I reverted that.

File: super_editor/test_goldens/super_textfield/super_textfield_font_height_test.dart (hunk 6 above)
Comment (angelosilvestre) [reply super_editor/test_goldens/super_textfield/super_textfield_font_height_test.dart]: Updated.

File: super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart (hunk 7 above)
Comment (angelosilvestre) [reply super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart:1263]: Can we make this change in a separate PR, so we are able to cherry-pick this PR to stable?

Comment (matthew-carroll) [reply super_editor/lib/src/super_textfield/infrastructure/text_scrollview.dart:1263]: Sure.

**Review by matthew-carroll — APPROVED** [review]
LGTM
